prometheus/compatible.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.  // You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+// compatibleHTTPClient must be used to contact a distant prometheus with a version >= v2.15
+type compatibleHTTPClient struct {
+	Client
+	requestTimeout   time.Duration
+	prometheusClient v1.API
+}
+
+func (c *compatibleHTTPClient) Metadata(metric string) (v1.Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
+	defer cancel()
+
+	metadata, err := c.prometheusClient.Metadata(ctx, metric, "1")
+	if err != nil {
+		return v1.Metadata{}, err
+	}
+	if len(metadata) == 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: metadata[metric][0].Type,
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+func (c *compatibleHTTPClient) AllMetadata() (map[string][]v1.Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
+	defer cancel()
+	return c.prometheusClient.Metadata(ctx, "", "")
+}
+
+func (c *compatibleHTTPClient) LabelNames(name string) ([]string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
```

Comment:
Is there a reason you're not propagating the contexts from the LSP request?

Since the LSP supports cancellation signals, those are actually of some use.

Reply:
When I wrote it, I was thinking about the client that won't use the network. And I thought it was ok to not force the usage of a context.

But then since it will be called from the LSP method with a context even with a headless server, then it doesn't make any sense.

I'll change that, thanks for the highlight :)

Reply:
done